internal/xds/xdsdependencymanager/xds_dependency_manager.go

```diff
@@ -0,0 +1,241 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package xdsdepmgr contains functions, structs, and utilities for working with
+// xds dependency manager. It will handle all the xds resources like listener,
+// route, cluster and endpoint resources and their dependencies.
+package xdsdepmgr
+
+import (
+	"context"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/pretty"
+	"google.golang.org/grpc/internal/status"
+	"google.golang.org/grpc/internal/xds/xdsclient"
+	"google.golang.org/grpc/internal/xds/xdsclient/xdsresource"
+)
+
+// DependencyManager struct stores all the information needed for watching all
+// the xds resources.
+type DependencyManager struct {
+	xdsClient xdsclient.XDSClient
+
+	watcher ConfigWatcher
+	// All methods on the xdsResolver type except for the ones invoked by gRPC,
+	// i.e ResolveNow() and Close(), are guaranteed to execute in the context of
+	// this serializer's callback. And since the serializer guarantees mutual
+	// exclusion among these callbacks, we can get by without any mutexes to
+	// access all of the below defined state. The only exception is Close(),
+	// which does access some of this shared state, but it does so after
+	// cancelling the context passed to the serializer.
+	serializer       *grpcsync.CallbackSerializer
+	serializerCancel context.CancelFunc
+
+	// dataplaneAuthority is the authority used for the data plane connections,
+	// which is also used to select the VirtualHost within the xDS
+	// RouteConfiguration.  This is %-encoded to match with VirtualHost Domain
+	// in xDS RouteConfiguration.
+	dataplaneAuthority string
+
+	ldsResourceName       string
+	listenerWatcher       *listenerWatcher
+	currentListenerUpdate *xdsresource.ListenerUpdate
+
+	rdsResourceName    string
+	currentRouteConfig xdsresource.RouteConfigUpdate
+	routeConfigWatcher *routeConfigWatcher
+	currentVirtualHost *xdsresource.VirtualHost
+
+	logger *grpclog.PrefixLogger
+}
+
+// New creates a new dependency manager that manages all the xds resources
+// including LDS, RDS, CDS and EDS and sends update once we have all the
+// resources and sends an error when we get error in listener or route
+// resources.
+func New(listenername, dataplaneAuthority string, xdsClient xdsclient.XDSClient, watcher ConfigWatcher) *DependencyManager {
+	// Builds the dependency manager and starts the listener watch.
```

Comment:
Nit: nix this comment as this is very obvious that you are creating the struct here. And the listener watch is not started here though.